fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,639 @@
+                                          remaining_ffg_weight: int) -> int:
+    assert min_ffg_support_slash_th <= ffg_support_for_checkpoint
+    
+    """
+    Return the max possible ffg_confirmation_score such that:
+    2 / 3 * total_active_balance <= \
+        ffg_support_for_checkpoint - \
+        min(min_ffg_support_slash_th, total_active_balance * ffg_confirmation_score / 100) + \
+        (1 - ffg_confirmation_score / 100) * remaining_ffg_weight
+    """
+
+    # Case 1: min_ffg_support_slash_th <= total_active_balance * ffg_confirmation_score / 100
+    #
+    # In this case the problem reduces to returning the max possible ffg_confirmation_score such that:
+    #
+    # 2 / 3 * total_active_balance <= \
+    #     ffg_support_for_checkpoint - min_ffg_support_slash_th + (1 - ffg_confirmation_score / 100) * 
+    #     remaining_ffg_weight
+    #
+    # If remaining_ffg_weight > 0, then first we compute ffg_confirmation_score and then check that the condition for
+    # this case is satisfied. If it is, then we return the calculated ffg_confirmation_score. If not, we proceed to the
+    # next case.
+    #
+    # If remaining_ffg_weight == 0, then the problem further reduces to whether the following condition, which is
+    # independent of ffg_confirmation_score, is satisfied:
+    #
+    # 2 / 3 * total_active_balance <= ffg_support_for_checkpoint - min_ffg_support_slash_th
+    #
+    # If it is satisfied, then it means that ffg_support_for_checkpoint can be as high as the maximum fraction of
+    # adversarial stake that the consensus protocol can deal with, i.e., < 1/3.
+    # If it not satisfied, then we proceed to the next case.
+
+    if remaining_ffg_weight > 0:
+        ffg_confirmation_score = (
+            100 * (
+                3 * (ffg_support_for_checkpoint - min_ffg_support_slash_th + remaining_ffg_weight) 
+                - 2 * total_active_balance
+            ) // (3 * remaining_ffg_weight)
+        )
+        if min_ffg_support_slash_th * 100 <= total_active_balance * ffg_confirmation_score:
+            return ffg_confirmation_score
+    else:
+        if 2 * total_active_balance <= (ffg_support_for_checkpoint - min_ffg_support_slash_th) * 3:
+            return 100 // 3
+
+    # Case 2: min_ffg_support_slash_th > total_active_balance * ffg_confirmation_score / 100
+    #
+    # In this case the problem reduces to returning the max possible ffg_confirmation_score such that:
+    #
+    # 2 / 3 * total_active_balance <= \
+    #     ffg_support_for_checkpoint - \
+    #     total_active_balance * ffg_confirmation_score / 100 + \
+    #     (1 - ffg_confirmation_score / 100) * remaining_ffg_weight
+
+    ffg_confirmation_score = (
+        100 * (3 * (ffg_support_for_checkpoint + remaining_ffg_weight) - 2 * total_active_balance) //
+        (3 * (remaining_ffg_weight + total_active_balance))
+    )
+    assert ffg_support_for_checkpoint * 100 > total_active_balance * ffg_confirmation_score
```

Comment:
Why would we need an assertion here? Would that mean that the computation is incorrect, in which case it can happen?

Reply:
If this assert is triggered, it would mean that there is a bug somewhere.